### PR TITLE
Fix navigation crash

### PR DESCRIPTION
**Pull request by vargaat**
### What happened?
- There were two crashes related to a navigation controller being pushed into another navigation controller.
- One crash was because of interactions between UIPrint* classes and based on the crash logs they no longer appear on iOS 18 so most probably this is an iOS level bug that got fixed.
- The other variant was most probably from our code but I couldn't reproduce the error not even based on the routes the user visited.
- To avoid such crashes I added a safeguard not to allow our nav controller to push another nav controller.
- I refactored our error / breadcrumb reporting to Crashlytics. My main concern there was that developer related log and error functions were mixed with the regular usage statistics methods and it was hard to understand what goes to which service. I separated developer related debug helping reporting methods into a dedicated entity to make things more easy. There are many code changes but most of them just about moving code around and renaming.
- I also changed when a new route is reported as a breadcrumb. Previously it was after presentation and in case the presentation crashed we lost the log. Now the route is logged before the actual routing takes place.

## Checklist

- [ ] Tested on phone
- [ ] Tested on tablet

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student, Teacher, Parent

[MBL-18060](https://instructure.atlassian.net/browse/MBL-18060)

<details><summary>:x: XCTest failed: CoreTests/StudioVideoPosterInteractorLiveTests/testReportsUnknownErrors</summary>


    XCTAssertEqual failed: ("0") is not equal to ("1")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("Studio Offline Sync Failed")")
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("random error")")
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against daac96d2c93b49c124fd9e2f64efc601015ad331
</p>

**Review by suhaibabsi-inst — DISMISSED** [review]
Thanks @vargaat !

Code +1

P.S. I was not able to test this on Firebase for lack of permission to DebugView. 
Will be working on fixing that issue.

**Review by rh12 — COMMENTED** [review]
Code review + 1

Listed a few minor things below
(plus the renaming to `DeveloperAnalytics` -> `RemoteLogger` and maybe `Logger` -> `LocalLogger`)

File: Core/Core/Analytics/DeveloperAnalytics.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+/// This entity encapsulates methods that log debug data for developers to make crash / error debugging easier.
+public struct DeveloperAnalytics {
+    public static var shared: DeveloperAnalytics = DeveloperAnalytics()
+
+    public weak var handler: DeveloperAnalyticsHandler?
+
+    #if DEBUG
+    /// If this property is true, then in addition to reporting to the analytics handler, events are also printed to the console.
+    public var logToConsole = false
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Analytics/DeveloperAnalytics.swift:32]: Left is as-is based on our conversation.

File: Core/CoreTests/UIViews/CoreNavigationControllerTests.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import Core
+import SwiftUI
+import XCTest
+
+class CoreNavigationControllerTests: CoreTestCase {
+
+    func testPushNavigationControllerNotCrashes() {
+        let testee = CoreNavigationController()
+        testee.developerAnalytics.handler = developerAnalytics
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/CoreTests/UIViews/CoreNavigationControllerTests.swift:27]: Removed the assignment but left the property on CoreNavigationController based on our conversation.

**Review by rh12 — APPROVED** [review]
QA + 1
(checked with xcode build only, locally all is well)

**Review by suhaibabsi-inst — APPROVED** [review]
Thanks @vargaat !

**Pull request merged**